From 976972c7c63f12a32c7a302f53fa7ca2c008ac8d Mon Sep 17 00:00:00 2001
From: Nathan Chancellor <nathan@kernel.org>
Date: Fri, 6 May 2022 10:22:01 -0700
Subject: [PATCH 1/2] metadata: os: Fallback to /usr/lib/os-release

When building TuxMake in a clean Arch Linux chroot [1], the metadata
operating system tests fail.

This happens because /etc/os-release does not exist in the chroot:

  # ls /etc/os-release
  ls: cannot access '/etc/os-release': No such file or directory

  # ls /usr/lib/os-release
  /usr/lib/os-release

On a Arch Linux system, /etc/os-release is not owned by any package but
/usr/lib/os-release is a part of the filesystem package, which is
require to be installed as part of the base package.

  $ pacman -Qo /etc/os-release
  error: No package owns /etc/os-release

  $ pacman -Qo /usr/lib/os-release
  /usr/lib/os-release is owned by filesystem 2021.12.07-1

This was reported as a bug [2] but it was closed with no apparent follow
up discussion.

Per the os-release specification [3], applications should check for the
existence of /etc/os-release and fallback to /usr/lib/os-release if it
does not exist. While the specification says that /etc/os-release should
be a relative symlink to /usr/lib/os-release, there is no language that
states this is a requirement; it just has to be in one of those
locations.

Modify the metadata commands to test for the existence of
/etc/os-release first, falling back to /usr/lib/os-release. With this
change, all the tests pass in both a clean chroot and an OCI container
on Arch Linux.

[1]: https://wiki.archlinux.org/title/DeveloperWiki:Building_in_a_clean_chroot
[2]: https://bugs.archlinux.org/task/64974
[3]: https://www.freedesktop.org/software/systemd/man/os-release.html

Signed-off-by: Nathan Chancellor <nathan@kernel.org>
---
 tuxmake/metadata/os.ini | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/tuxmake/metadata/os.ini b/tuxmake/metadata/os.ini
index a6cec35..b072a0e 100644
--- a/tuxmake/metadata/os.ini
+++ b/tuxmake/metadata/os.ini
@@ -2,5 +2,5 @@
 order = 4
 
 [commands]
-name = . /etc/os-release && echo $NAME
-version = . /etc/os-release && (test -n "$VERSION_ID" && echo $VERSION_ID) || (test -n "$BUILD_ID" && echo $BUILD_ID) || cat /etc/debian_version
+name = test -e /etc/os-release && . /etc/os-release || . /usr/lib/os-release && echo $NAME
+version = test -e /etc/os-release && . /etc/os-release || . /usr/lib/os-release && (test -n "$VERSION_ID" && echo $VERSION_ID) || (test -n "$BUILD_ID" && echo $BUILD_ID) || cat /etc/debian_version
-- 
2.36.0

